DM-4950

jonathansick opened the pull request:
This PR establishes LTD Keeper, although the service is not fully complete and deployed through this ticket. This PR is to verify that the Flask underpinnings and API structure of the web app are on the right track.

See docs at http://ltd-keeper.lsst.io

Work deferred to other tickets:

- DM-5167 will hook up external interactions with AWS and Fastly,
- DM-5291 will improve the configuration system, moving it towards granular environment variables
- DM-5194 will dockerize LTD Keeper.

jonathansick commented:
FYI the Read the Docs doesn't seem to be serving the latest version of the docs at http://ltd-keeper.lsst.io/en/tickets-dm-4950/ (caching issue on their side?). I'll post when the API docs are current.

jmatt commented on LICENSE:
```diff
@@ -19,3 +19,29 @@ AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+
+The MIT License (MIT)
+
+Copyright (c) 2014 Miguel Grinberg
```
Copyright is wrong.

jonathansick replied:
What's the correct way to do this?

timj replied:
You can't delete the copyright line. You can only add to it. You may want to add a sentence distinguishing your code from Miguel's code. Was the original code added as a distinct commit?

jonathansick replied:
The whole app architecture is derived from https://github.com/miguelgrinberg/oreilly-flask-apis-video but it wasn't a fork. I essentially rebuilt the app with his structure commit-by-commit (and mention that in the commit messages).

jonathansick replied:
@jmatt Can you please clarify exactly what I should be doing here?

timj replied:
I just spoke to @jonathansick about it so I think it's under control now.

jmatt commented on README.rst:
```diff
@@ -10,8 +10,9 @@ Keeper is implemented as a Python `Flask <http://flask.pocoo.org>`_ app.
 
```
Make sure the README has a reference to http://ltd-keeper.lsst.io/.

jmatt commented on app/__init__.py:
```diff
@@ -0,0 +1,71 @@
+db = SQLAlchemy()
+
+
+def create_app(config_name):
```
Default `config_name` so it'll work even when instantiating the app through a different method. For instance using uwsgi instead of run.py.

jmatt commented on integration_tests/doc_example_driver.sh:
```diff
@@ -0,0 +1,142 @@
+# 1. Delete any existing ltd-keeper-dev.sqlite
+# 2. Run the dev server ./run.py
+# 3. Get an auth token
+#        http --auth user:pass GET http://localhost/token
```
This uses httpie. You should probably link to it here and the README.

https://github.com/jkbrzt/httpie

jmatt commented on integration_tests/doc_example_driver.sh:
```diff
@@ -0,0 +1,142 @@
+
+http -p hbHB --pretty=format --auth $KEEPER_TOKEN: \
+    DELETE http://localhost:5000/editions/1 \
+    >> ltd_keeper_doc_examples.txt
```
Unless there is a specific reason to put it in ./ltd_keeper_doc_examples.txt, I'd output this to stdout and let me redirect.

jmatt commented on app/api_v1/builds.py:
```diff
@@ -0,0 +1,313 @@
+    return jsonify(build.export_data()), 201, {'Location': build.get_url()}
+
+
+@api.route('/builds/<int:id>/uploaded', methods=['POST'])
```
This could be a PATCH on a build instead of a POST. Usually a POST implies creation or some sort of side effect. I.e. If it were actually uploading the build then that would be a POST.

jmatt commented on docs/api_intro.rst:
```diff
@@ -0,0 +1,44 @@
+
+The LTD Keeper API is in beta and can be changed without notice.
+In the future, we may elect to version the API through the ``Accept`` header.
+
```
I wouldn't commit to any specifics related to future versioning. Just call it v1 (or the initial version) and leave it at that.

I guess SQRE may really want to have an opinion on how to version future APIs, if that's the case it should be discussed.

jmatt commented on app/api_v1/builds.py:
```diff
@@ -0,0 +1,313 @@
+    """
+    build_urls = [build.get_url() for build in
+                  Build.query.filter(Product.slug == slug).all()]
+    return jsonify({'builds': build_urls})
```
Everything is technically correct with get_product_builds (/builds). But you are returning deprecated builds here too. I don't know if that'll be handled somewhere else? Or there could be more filtering here (possibly through parameters).

jonathansick replied:
:white_check_mark: I've added a filter on both `GET /products/<slug>/builds/` and `GET /products/<slug>/editions/` against these deprecated builds/editions. At the moment the individual resources can still be retrieved. I could 404 on this but if the user is specifically looking for something, the user will also see from the `date_ended` field that the record is deprecated. So... might be ok here.

jmatt commented:
Just a note that this PR does not cover the deploy. That'll be in a future PR.

jmatt commented:
SQR-006 and the [API introduction](http://ltd-keeper.lsst.io/en/tickets-dm-4950/api_intro.html) have some versioning discrepancies. References using `/v1/` in the URL.

jonathansick commented:
I think I've addressed all of @jmatt comments here. I've also added Travis testing.

I'd say things are good to merge this and pick up the smaller issues while deploying as alpha, eh?
